Include plaintext response body in `event`

aggroskater opened the pull request:
Presently, XML and JSON response bodies get encoded into the returned
`event` object. These bodies can then be used later in flows (for
filters, or for including data in other destinations). However,
plaintext bodies are not included in the `event` object.

This leads to using success/failure/reason regex in the mappings
themselves to interrogate the response body, as that is presently the
only place where it can be examined. In this manner, one can effectively
"inject" the body (or some portion of it) into the "reason" field. This
also necessarily modifies the "outcome", even if that might not be the
desired effect.

An example would be an enhancement that returns "true" or "false"
depending on whether an email address is a duplicate. In either case,
the "outcome" for the enhancement could be considered "success" --that
is, we only want the "outcome" to be an error or failure if the endpoint
itself failed-- and filters would then want to examine the response body
to take appropriate action. Another example would be an enhancement that
returns a single number as a score. Perhaps we'd like to include that in
a subsequent destination. Presently, since the response body is not
available, the success/failure/reason regex is the only available
mechanism for performing these actions.

This commit adjusts the response handler to include non-XML, non-JSON,
non-HTML responses in a "body" property of the `event` object. This
permits plaintext responses to have the same extensibility as JSON and
XML responses. The commit also adjusts several tests that previously
were expecting just the "outcome" in the `event` object.

aggroskater commented:
Let me know what y'all think about this change. The present approach of using the success/failure/reason regex mappings "works", but doesn't feel like the proper way to do things, seeing as the enhancements themselves aren't "failing", per se.

cgrayson commented:
Thanks for the suggestion, and thanks even more for the implementation :-) and apologies for not responding sooner. "Injecting" the body into `reason` is clever, but I agree, it isn't clean or straightforward. 

On the other hand, we're worried about adding the entire unstructured text of the body onto these events. In LC Classic, there's a 4k limit on how much of the response is retained, and it's not too uncommon to see that have to be enforced (usually for HTML page responses).

We also wondered about the usefulness of this data, especially in the context of rule-building, in the absence of a string-matching operator. We plan to add that, but it's not available yet. As I re-read your examples, however, they seemed as though they might mainly be single-word strings: "true", "false", a single number as a score. And for plain text, as opposed to HTML, say, that does seem common. 

All of which to say, I wonder if some kind of limiting or truncation would be able to support your cases (and a majority of similar cases), but protect LC from storing inordinate amounts of less-than-useful data. A simple max-character limit would be one approach, or only capturing up to the first whitespace, for example. Thoughts?

aggroskater commented:
Truncation would likely work for our cases. I've added a commit to limit to the first 128 characters. I also added a typeof sanity check so I don't end up calling .substring on something that isn't a string. Finally, there's an added test case to make sure we only store the first 128 characters. This change should also be amenable to any future plans for string matching operators (use the captured string if available, otherwise, default to first 128 characters of response).

Sidenote: the code tries to make sure we aren't capturing HTML to begin with, but of course, if the response body is malformed --malformed JSON, XML, HTML, whatever-- it's definitely possible that the string fallback would be filled with many kibibytes of less-than-useful data.
